vaadin-dialog-flow-parent/vaadin-dialog-flow/src/main/java/com/vaadin/flow/component/dialog/Dialog.java

```diff
@@ -681,13 +683,49 @@ public void remove(Component... components) {
         /**
          * Removes all components from the container.
          */
+        @Override
         public void removeAll() {
             root.removeAllChildren();
             dialog.getElement()
                     .executeJs("this." + rendererFunction + " = null;");
             setRendererCreated(false);
         }
 
+        /**
+         * Adds the given component to the container at the specific index.
+         *
+         * @param index
+         *            the index, where the component will be added. The index
+         *            must be non-negative and may not exceed the children count
+         *            component – the component to add, value should not be null
+         * @param component
+         *            the component to be added.
+         */
+        @Override
+        public void addComponentAtIndex(int index, Component component) {
+            Objects.requireNonNull(component, "Component should not be null");
+            if (index < 0) {
+                throw new IllegalArgumentException(
+                        "Cannot add a component with a negative index");
+            } else {
+                root.insertChild(index, component.getElement());
+            }
+            if (!isRendererCreated()) {
+                initRenderer();
+            }
```

Comment:
There will be quite a bit of duplication of this. It looks like we could introduce a single method for all APIs, that both sets or removes the renderer:
- It checks if the element has children or not
- If it does, it sets the renderer unless it already exists
- If not, it removes the renderer

Reply:
Extracted the logic to `updateRendererState` and it is called in all overridden methods.